src/freenet/client/async/SimpleHealingQueue.java

```diff
@@ -73,6 +81,57 @@ public boolean innerQueue(Bucket data, byte[] cryptoKey, byte cryptoAlgorithm, C
 		}
 	}
 
+	/**
+	 * Specialize Healing to the fraction of the keyspace in which we would receive the inserts
+	 * if we were one of 5 long distance nodes of an actual inserter.
+	 *
+	 * If an opennet node is connected to an attacker, healing traffic could be mistaken for an insert.
+	 * Since opennet cannot be fully secured, this should be avoided.
+	 * As a solution, we specialize healing inserts to the inserts we would send if we were one of 5 
+	 * long distance connections for a node in another part of the keyspace.
+	 *
+	 * As a welcome side effect, specialized healing inserts should take one hop less to reach the
+	 * correct node from which loop detection will stop the insert long before HTL reaches zero.
+	 */
+	private boolean isHealingThisBlockSimilarToForwarding(
+			ClientContext context,
+			SingleBlockInserter sbi) {
+		// pure darknet is safer against sybil attacks, so we can heal fully
+		if (!opennetEnabled) {
+			return true;
+		}
+		// ensure that we have a routing key
+		sbi.tryEncode(context);
+		double keyLocation = sbi.getKeyNoEncode().getNodeKey().toNormalizedDouble();
+		double randomBetweenZeroAndOne = NodeStarter.getGlobalSecureRandom().nextDouble() / Double.MAX_VALUE;
+		return shouldHealBlock(nodeLocation, keyLocation, randomBetweenZeroAndOne);
+	}
+
+	/**
+	 * choose fraction by probabilistic dropping of far away keys.
+	 * only enqueue keys in our 20% of the keyspace: the ones that would reach us if we were one of
+	 * 5 long distance peers of a peer node.
+	 */
+	static boolean shouldHealBlock(
+			double nodeLocation,
+			double keyLocation,
+			double randomBetweenZeroAndOne) {
+		double distanceToNodeLocation = Location.distance(nodeLocation, keyLocation);
+		// accept half the healings in our 20% of the keyspace.
+		// If the key is inside "our" 20% of the keyspace, heal it with 50% probability.
+		if (distanceToNodeLocation < 0.1) {
+			// accept 50%, specialized to our own location (0.5 ** 4 ~ 0.0625). Accept 70% which are going
```

Comment:
I spent several minutes mulling this over, and either this comment has nothing to do with is happening here, or something is very off. 😁

Reply:
This should actually be somewhat correct, but there are several jumps in abstraction layers and logic between the comment and the implementation.  I think this needs more prose in the javadoc to actually explain what it does.